internal/components/datasource/coremodel.go

```diff
@@ -1,190 +1,39 @@
 package datasource
 
 import (
-	"context"
-	"errors"
-	"time"
-
-	kerrors "k8s.io/apimachinery/pkg/api/errors"
-	ctrl "sigs.k8s.io/controller-runtime"
-	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/reconcile"
-
-	"github.com/grafana/grafana/internal/components"
-	"github.com/grafana/grafana/pkg/infra/log"
-	"github.com/grafana/grafana/pkg/models"
-	"github.com/grafana/grafana/pkg/schema"
+	"github.com/grafana/thema"
 )
 
-// Coremodel is the coremodel for Datasource component.
+// Coremodel contains the foundational schema declaration for datasources.
 type Coremodel struct {
-	schema schema.ObjectSchema
-	store  components.Store
-	client client.Client
-	logger log.Logger
+	lin thema.Lineage
 }
 
-// ProvideCoremodel provides a new Coremodel with store and schema loaded from loader.
-//
-// TODO: this is currently done manually and is statically enumerated in the registry.
-// We should figure out a way to dynamically register this to registry and automate schema loading too,
-// since the loading process will be exactly the same for all components (except for schema options).
-func ProvideCoremodel(store components.Store, loader components.SchemaLoader) (*Coremodel, error) {
-	schema, err := loader.LoadSchema(context.TODO(), schema.SchemaTypeThema, schema.ThemaLoaderOpts{
-		SchemaFS:         cueFS,
-		SchemaPath:       cuePath,
-		SchemaVersion:    schemaVersion,
-		GroupName:        groupName,
-		GroupVersion:     groupVersion,
-		SchemaOpenapi:    schemaOpenapi,
-		SchemaType:       &DatasourceSpec{},
-		SchemaObject:     &Datasource{},
-		SchemaListObject: &DatasourceList{},
-	}, schema.GoLoaderOpts{})
-	if err != nil {
-		return nil, err
-	}
-
-	return &Coremodel{
-		store:  store,
-		schema: schema,
-		logger: log.New("components.datasource.coremodel"),
-	}, nil
+func (c *Coremodel) Lineage() thema.Lineage {
+	return c.lin
 }
 
-// Schema returns the object schema for this Coremodel.
-func (m *Coremodel) Schema() schema.ObjectSchema {
-	return m.schema
+func (c *Coremodel) Schema() thema.Schema {
+	sch, err := c.lin.Schema(currentVersion)
+	if err != nil {
+		// Only reachable if our own schema currentVersion does not exist, which
+		// can really only happen transitionally during development
+		panic(err)
+	}
+	return sch
```

Comment:
I'm not a fan of this approach. Since `Lineage#Schema` returns an error, can we inject `thema.Schema` to `Coremodel` during loading, based on the version, instead of reading it here, because there's no way of dealing with errors?

Alternatively, we could change the signature of the method to return an error.

Reply:
It's plausible that this whole "Thema + current Go version" gets moved generically into Thema, anyway - it's already sorta there with `InputKernel`

Reply:
I very much dislike panics in general, especially ones that happen at runtime due to misconfiguration. I think it's OK to leave it as it is right now, but I'd rather have it panic at build time or gracefully handle the error (e.g. skip loading malformed schema, or at least exit gracefully).